automation/routines.go

```diff
@@ -0,0 +1,170 @@
+	// Early out if there is nothing to filter.
+	if len(routines) == 0 {
+		return routines
+	}
+
+	// Construct a nested lookup map without duplicates, or early out on a match
+	// all condition.
+	//
+	// The map uses the first element of the path as a key. As a special case
+	// "*" will match all.
+	var matchAll bool
+	lookup := make(map[string][]string, len(patterns))
+LOOKUP:
+	for _, path := range patterns {
+		name, nestedPath, _ := strings.Cut(path, "/")
+
+		var found bool
+		if name == "*" {
+			found = true
+		} else {
+			_, found = routines[name]
+		}
+
+		switch {
+		case !found:
+			// Entry not found; nothing to do.
+		case name == "*" && nestedPath == "":
+			// Match all or end of path; early out.
+			matchAll = true
+			break LOOKUP
+			// Routine not found; continue.
+		case len(lookup[name]) == 1 && lookup[name][0] == "":
+			// Path already match with an end of-path criteria; nothing to do.
+		case nestedPath == "":
+			// End of path; replace existing lookup as the end-of-path criteria
+			// match all cases.
+			lookup[name] = []string{""}
+		case len(lookup[name]) == 1 && lookup[name][0] == "*":
+			// Path already match all sub-routines; nothing to do.
+		case nestedPath == "*":
+			// Match all sub-routines; replace existing lookup with a wildcard
+			// criteria.
+			lookup[name] = []string{"*"}
+		default:
+			// Append nested lookup path.
+			lookup[name] = append(lookup[name], nestedPath)
+		}
+	}
+
+	if matchAll {
+		return routines
+	}
+
+	// Filter routines based on the lookup map.
+	filtered := make(Routines, len(patterns))
+	var nestedPath []string
+	for name, r := range routines {
+		// Reset subPatterns before use.
+		nestedPath = nestedPath[:0]
+		// Add all patterns that apply to name.
+		nestedPath = append(nestedPath, lookup["*"]...)
+		nestedPath = append(nestedPath, lookup[name]...)
+
+		slices.Sort(nestedPath)
+		rs, canNest := r.(Routines)
+		switch {
+		case len(nestedPath) == 0:
+			// No lookup matching the routine; skip entry.
+		case slices.Contains(nestedPath, ""):
+			// End of path; add routine.
+			filtered[name] = r
+		case !canNest:
+			// Remaining matchers require the canNest property.
+		case slices.Contains(nestedPath, "*"):
+			// Match all sub-routines.
+			filtered[name] = rs
+		default:
+			// Match named sub-routines.
+			filtered[name] = rs.SubRoutines(nestedPath...)
+		}
+	}
+
+	return filtered
+}
+
+// Do runs the member routines in an alphanumerical order and assigns correct
+// sub-routine names. If cfg.EarlyOut() returns true, return at the first error.
+// Otherwise log the error and continue.
+func (routines Routines) Do(ctx context.Context, cfg *Config) error {
+	earlyOut := cfg.EarlyOut()
+
+	keys := make([]string, 0, len(routines))
+	for k := range routines {
+		keys = append(keys, k)
+	}
+	slices.Sort(keys)
+
+	var errCnt int
+	for _, k := range keys {
+		r := routines[k]
+		cfg := cfg.WithSubRoutineName(k)
+		logger := cfg.Logger()
```

Comment:
why start a brand new logger for every routine?

Reply:
Because we want the sub-routine name to be injected before we log.

Reply:
can't we use `.With()` from that initial logger?

Reply:
No, we can't. We are not allowed to _replace_ existing keys with the `With` statement, so for nested routines, we would get key duplicates: https://go.dev/play/p/n2Sre6ofAPb

```go
func main() {
	l := slog.Default().With("key1", "folder1")
	l.Info("test1")
	l = l.With("key1", "folder1/task1")
	l.Info("test2")
}
```
```txt
2009/11/10 23:00:00 INFO test1 key1=folder1
2009/11/10 23:00:00 INFO test2 key1=folder1 key1=folder1/task1
```

However we are not creating a new logger from scratch here; the cfg contains a base logger without leaf fields. When we call cfg.Logger(), it will fork of the base logger and append the _correct_ leaf fields to the logger. This is the same stuff that we need to do for e.g. zerolog, so it should be considered a common enough pattern.

So to summarize:
1. For each sub-routine we _fork_ the cfg from the base cfg and appends the current routine name.
2. We get a logger from cfg which _forks_ the cfg base logger and add leaf fields (including routine name, which is the one we altered).

Reply:
you don't need to override the parent logger and instead for each routine create a child: https://go.dev/play/p/I_lKk2HIR8Q

```
func main() {
	parent := slog.Default()
	parent.Info("test_parent")

	child1 := parent.With("key1", "folder1/task1")
	child1.Info("test_child1")

	child2 := parent.With("key1", "folder1/task2")
	child2.Info("test_child2")
}
```

```
2009/11/10 23:00:00 INFO test_parent
2009/11/10 23:00:00 INFO test_child1 key1=folder1/task1
2009/11/10 23:00:00 INFO test_child2 key1=folder1/task2
```

Reply:
Anyway, feel free to stick to your solution 👍

Reply:
@sarzhann  this is what we are already doing, it's just that it happens _inside_ the `cfg.Logger()`  function which is responsible for guarding the parent logger.